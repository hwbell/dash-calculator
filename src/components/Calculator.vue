<template>
  <div class="calculator">
    <h3>{{ msg }}</h3>

    <ExpressionWindow
      v-bind:expression="expressionDisplay"
      v-bind:answer="currentAnswer"
      v-bind:alert="alertSwitch"
    />

    <div class="calculator-buttons">
      <div class="" v-for="(buttonrow, ind) in buttons" v-bind:key="ind">
        <button
          v-for="(button, ind) in buttonrow"
          v-bind:key="ind"
          class="calculator-button btn"
          :class="!!button.class ? 'btn-' + button.class : 'btn-light'"
          v-on:click="modifyExpression(button)"
        >
          {{ button.text }}
        </button>
      </div>
    </div>
  </div>
</template>

<script>
import ExpressionWindow from "./ExpressionWindow";

export default {
  name: "Calculator",
  props: {
    msg: String,
  },
  components: { ExpressionWindow },
  data: function() {
    return {
      // the currentExpressionText and currentExpressionMath below will generally look the same, but must be held differently for user readability(text) and eval() usability for computing the answer (math).
      currentExpressionText: [],
      currentExpressionMath: [],
      // miniExpression: "",
      currentAnswer: null,
      // the alertSwitch is to turn the text red on an invalid expression.
      alertSwitch: false,
      buttons: [
        [
          {
            text: " (",
            math: " (",
            class: "secondary",
          },
          {
            text: ") ",
            math: ") ",
            class: "secondary",
          },
          {
            text: " % ",
            math: "/100",
            class: "secondary",
          },
          {
            text: "CE",
            class: "secondary",
          },
        ],
        [
          {
            text: "7",
            math: "7",
          },
          {
            text: "8",
            math: "8",
          },
          {
            text: "9",
            math: "9",
          },
          {
            text: " / ",
            math: " / ",
            class: "secondary",
          },
        ],
        [
          {
            text: "4",
            math: "4",
          },
          {
            text: "5",
            math: "5",
          },
          {
            text: "6",
            math: "6",
          },
          {
            text: " x ",
            math: " * ",
            class: "secondary",
          },
        ],
        [
          {
            text: "1",
            math: "1",
          },
          {
            text: "2",
            math: "2",
          },
          {
            text: "3",
            math: "3",
          },
          {
            text: " - ",
            math: " - ",
            class: "secondary",
          },
        ],
        [
          {
            text: "0",
            math: "0",
          },
          {
            text: ".",
            math: ".",
          },
          {
            text: "=",
            class: "primary",
          },
          {
            text: " + ",
            math: " + ",
            class: "secondary",
          },
        ],
      ],
    };
  },
  methods: {
    modifyExpression: function(button) {
      // `this` inside methods point to the Vue instance

      if (button.text === "=") {
        // evaluate
        this.evaluateExpression();
      } else if (button.text === "CE") {
        // split by spaces to properly chop the last button's contribution
        this.currentExpressionText.pop();
        this.currentExpressionMath.pop();
        this.currentAnswer = null;
      } else {
        // add a space for chopping above
        this.currentExpressionText.push(button.text);
        this.currentExpressionMath.push(button.math);
      }
    },
    evaluateExpression: function() {
      try {
        let expressionString = this.currentExpressionMath.join("");
        this.currentAnswer = eval(expressionString);
      } catch(e) {
        console.log(e);
        this.alertSwitch = true;
        setTimeout(() => {
          this.alertSwitch = false;
        }, 300)
      }
    },
  },
  computed: {
    // a computed getter
    expressionDisplay: function() {
      // `this` points to the vm instance
      return this.currentExpressionText.join("");
    },
  },
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped>
h3 {
  margin: 40px;
}
.calculator {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}
.calculator-buttons {
  width: 80%;
}
.calculator-button {
  padding: 10px;
  width: 100px;
  border-radius: 0;
  border: 1px solid rgb(255, 255, 255);
}
</style>
