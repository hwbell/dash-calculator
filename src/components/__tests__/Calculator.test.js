import { mount } from '@vue/test-utils'
import Calculator from '../Calculator.vue'

describe('Calculator', () => {

  let wrapper;
  beforeEach(() => {
    wrapper = mount(Calculator)
  })

  test('is a Vue instance', () => {
    expect(wrapper).toBeTruthy()
  })

  test('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
  
  test('renders a button for each object in the array map', () => {
    expect(wrapper.findAll('button').length).toBe(20);

    // test a unique button(s)
    let button = wrapper.find('.btn-primary');
    expect(button).toBeTruthy();
    expect(button.text()).toBe('=')

    // test a group of button(s)
    let secondaryButtons = wrapper.findAll('.btn-secondary');
    expect(secondaryButtons.length).toBe(8);
    expect(secondaryButtons.at(0).text()).toBe('(')
    
  })

  test(`should register the button's math value with the expression state`, () => {
    // things start out empty
    expect(wrapper.vm.currentExpressionText).toEqual([]);
    expect(wrapper.vm.currentExpressionMath).toEqual([]);

    // and then when we click some buttons ...
    wrapper.find('#calculator-button-0-0').trigger('click')
    wrapper.find('#calculator-button-1-2').trigger('click')
    wrapper.find('#calculator-button-3-0').trigger('click')
    wrapper.find('#calculator-button-0-1').trigger('click')
    expect(wrapper.vm.currentExpressionText).toEqual([' (', '9', '1', ') ']);
    expect(wrapper.vm.currentExpressionMath).toEqual([' (', '9', '1', ') ']);

    // reset
    wrapper.vm.currentExpressionMath = [];
    wrapper.vm.currentExpressionText = [];

    // try one that creates different arrays for math / text
    wrapper.find('#calculator-button-1-2').trigger('click')
    wrapper.find('#calculator-button-2-2').trigger('click')
    wrapper.find('#calculator-button-0-2').trigger('click')
    wrapper.find('#calculator-button-2-3').trigger('click')
    expect(wrapper.vm.currentExpressionText).toEqual(['9', '6', ' % ', ' x ']);
    expect(wrapper.vm.currentExpressionMath).toEqual(['9', '6', '/100', ' * ']);

  })

  test('should evaluate a valid expression', () => {
    // starts out null
    expect(wrapper.vm.currentAnswer).toBe(null);

    // and then when we click some buttons ...
    wrapper.find('#calculator-button-1-0').trigger('click')
    wrapper.find('#calculator-button-1-1').trigger('click')
    wrapper.find('#calculator-button-1-2').trigger('click')
    wrapper.find('#calculator-button-2-3').trigger('click')
    wrapper.find('#calculator-button-2-2').trigger('click')
    wrapper.find('#calculator-button-1-2').trigger('click')
    wrapper.find('#calculator-button-1-1').trigger('click')
    
    expect(wrapper.vm.currentExpressionText).toEqual(['7', '8', '9', ' x ', '6','9','8']);
    expect(wrapper.vm.currentExpressionMath).toEqual(['7', '8', '9', ' * ', '6','9','8']);

    // evaluate
    wrapper.find('#calculator-button-4-2').trigger('click')
    expect(wrapper.vm.currentAnswer).toBe(550722);

  })
})