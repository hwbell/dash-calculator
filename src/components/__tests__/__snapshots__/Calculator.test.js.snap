// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator renders correctly 1`] = `
<div
  class="calculator"
>
  <h4>
    Calculator
  </h4>
   
  <div
    class="expression-window"
  >
    <div
      class="expression-text"
    >
      0Â 
    </div>
     
    <!---->
  </div>
   
  <div
    class="calculator-buttons"
  >
    <div>
      <button
        class="calculator-button btn btn-secondary"
      >
        
         (
      
      </button>
      <button
        class="calculator-button btn btn-secondary"
      >
        
        ) 
      
      </button>
      <button
        class="calculator-button btn btn-secondary"
      >
        
         % 
      
      </button>
      <button
        class="calculator-button btn btn-secondary"
      >
        
        CE
      
      </button>
    </div>
    <div>
      <button
        class="calculator-button btn btn-light"
      >
        
        7
      
      </button>
      <button
        class="calculator-button btn btn-light"
      >
        
        8
      
      </button>
      <button
        class="calculator-button btn btn-light"
      >
        
        9
      
      </button>
      <button
        class="calculator-button btn btn-secondary"
      >
        
         / 
      
      </button>
    </div>
    <div>
      <button
        class="calculator-button btn btn-light"
      >
        
        4
      
      </button>
      <button
        class="calculator-button btn btn-light"
      >
        
        5
      
      </button>
      <button
        class="calculator-button btn btn-light"
      >
        
        6
      
      </button>
      <button
        class="calculator-button btn btn-secondary"
      >
        
         x 
      
      </button>
    </div>
    <div>
      <button
        class="calculator-button btn btn-light"
      >
        
        1
      
      </button>
      <button
        class="calculator-button btn btn-light"
      >
        
        2
      
      </button>
      <button
        class="calculator-button btn btn-light"
      >
        
        3
      
      </button>
      <button
        class="calculator-button btn btn-secondary"
      >
        
         - 
      
      </button>
    </div>
    <div>
      <button
        class="calculator-button btn btn-light"
      >
        
        0
      
      </button>
      <button
        class="calculator-button btn btn-light"
      >
        
        .
      
      </button>
      <button
        class="calculator-button btn btn-primary"
      >
        
        =
      
      </button>
      <button
        class="calculator-button btn btn-secondary"
      >
        
         + 
      
      </button>
    </div>
  </div>
</div>
`;
